<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title>Mutation of reads</title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
  <link rel="stylesheet" href="manual.css" type="text/css" />
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
</head>
<!-- BODY BEGIN ------------------------------>
<body>
<div class="toc contentbox">
    <span class="tocheader">Table of contents</span>
<ul>
<li><a href="index.html">Package overview</a></li>
<li><a href="00b_shell-prerequisites.html">Dependencies of the scripts</a>
<li><a href="00c_input-files.html">Input files</a>
<li><a href="01_genome-preparation.html">Genome Preparation</a>
<li><a href="02_read-generation.html">Generation of simulated read sets</a>
<li class="tocthis"><a href="03_read-mutation.html">Mutation of reads</a>
<li><a href="04_mapping.html">Generate mapper calls</a>
<li><a href="05_writing-a-mapper-script.html">Writing a mapper script</a>
<li><a href="06_sam-parsing.html">Parsing of SAM files</a>
<li><a href="07_parameter_influence.html">Comparing multiple mapper runs</a>
<li><a href="A01_workflow.html">Appendix A1 -- Tutorial script</a>
<li><a href="A02_advanced_read-generation.html">Appendix A2 -- Advanced read generation</a>
<li><a href="A03_scripts.html">Appendix A3 -- Overview of the scripts by category</a>
</ul>
</div>
<div class="content contentbox">
<div id="header">
<h1 class="title">Mutation of reads</h1>
</div>
<!-- Local Table of contents -->
<!--Back to table of contents -->
<!-- CONTENT -------------------------------------->
<p>Mutations must be introduced into the artificial reads in order to simulate chemical damage and evolutionary distance of <em>ancient DNA</em> with respect to the recent reference which they shall be mapped to.</p>
<p>As chemical damage is more prevalent near the end of <em>ancient DNA</em>, TAPAS can introduce more mutations into the artificial reads, near their ends. To this effect parameters must be determined, which affect the probability of each base exchange for each position in the read.</p>
<p>The mutation probability parameters are detailed in the <a href="#mutation-probabilities">next section</a>. Following that, the section <a href="#specifying-the-parameters">Specifying the parameters</a> shows how values for the mutation probability parameters can be set by the user. Afterwards, in <a href="#subject-a-fastq-file-to-artificial-mutations">Subject-a-FASTQ-file-to-artificial-mutations</a>, the FASTQ file containing artificial endogenous reads, which was generated in the <a href="02_read-generation.html">previous topic</a>, is subjected to mutations with the specified base exchange probabilities.</p>
<p>This chapter closes with two <a href="#advanced-read-mutation">advanced topics</a>: Obtaining mutation probability parameters from a mapDamage output and how to define multiple sets of mutation probability parameters to create multiple FASTQ files containing reads of different mutation strengths.</p>
<h2 id="mutation-probabilities">Mutation probabilities</h2>
<p>The reads are mutated using per base probabilities derived from the geometric distribution. The mutation probability at the read ends is the highest. By this, the chemical damage near read ends can be modelled. For this, three parameters are important:</p>
<ul class="incremental">
<li><p>The position-independent mutation probability <span class="math inline">\(t\)</span> (<span class="math inline">\(0&lt;t&lt;1\)</span>). This is the probability of any base to mutate, regardless of its proximity to the end of the read. This can be used to model evolutionary distance.</p></li>
<li><p>The steepness <span class="math inline">\(p\)</span> (<span class="math inline">\(0&lt;p&lt;1\)</span>) of the mutation probability decline when moving away from the read end. The higher this parameter the steeper is the decline of mutation probability when moving away from the read end.</p></li>
<li><p>The multiplying factor <span class="math inline">\(f\)</span>. At the read end, <span class="math inline">\(p\cdot f+t\)</span> is the probability of the first base of the read to be mutated.</p></li>
</ul>
<p>With this model, its possible to archive mutation probabilities greater than one. This makes of course no sense and the mutation probability is cut back to one in such cases.</p>
<p>In mathematical notation, the mutation probability <span class="math inline">\(P_{\text{mut}}\)</span> of a base number <span class="math inline">\(x\)</span>, starting to count at the reads' end, is:</p>
<p><span class="math display">\[P_{\text{mut}}(x) = f \cdot \text{dgeom}(x;p) + t\]</span></p>
<p>with <span class="math inline">\(\text{dgeom}(X;P)\)</span> being the density function of the geometric distribution, with parameters X = number of tries; P = success probability.</p>
<p>The following sketch graphs illustrate the three parameters:</p>
<p><img src="fig/mut.svg" width="800" /></p>
<h2 id="specifying-the-parameters">Specifying the parameters</h2>
<p>The mutation probabilities are saved as a text table. It looks like this:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">cat</span> input/mut-tables/mut.tab</code></pre></div>
<pre class="output"><code>strand   from   to   factor  geom_prob  intercept
3p       C      T    0.3     0.4        0.1
5p       C      T    0.1     0.2        0.0
3p       *      *    0.0     0.1        0.12</code></pre>
<p>Two types of per-base mutation probabilities are distinguished:</p>
<p>The columns have the following meaning:</p>
<ul class="incremental">
<li>strand: [3 or 5] Which side (3' or 5' end) of the read shall be considered the read end when the base-dependent mutation probability is calculated. If both sides of the read shall see heightened mutation probabilities, include two lines, one with a value <code>3</code> and one with a value <code>5</code>.</li>
<li>from: [Letter or <code>*</code>] Which bases should mutate according to this lines' parameters. The character <code>*</code> means this line applies to every base.</li>
<li>to: [Letter or <code>*</code>] Which base should a mutation event yield. <code>*</code> means a base (A,T,C,G) different from the original base is chosen at random.</li>
<li>factor: Maximum position-dependent mutation probability at the read end (specified by <code>strand</code>)</li>
<li>geom_prob: How fast the position-dependent mutation probability declines for bases further away from the read end. 0 means no decline, higher values lead to faster decline.</li>
<li>intercept: The position-independent mutation probability.</li>
</ul>
<p>For each row of the table, the read is subjected to one round of mutation introduction where the mutation probability per base is determined by the parameters found in that row.</p>
<h2 id="subject-a-fastq-file-to-artificial-mutations">Subject a FASTQ file to artificial mutations</h2>
<p>The <code>multiple-mutate.py</code> tool takes a table of the previous section as input and mutates raw nucleotide strings provided to it on standard input accordingly.</p>
<p>An already existent FASTQ file can be mutated using the <code>filter_fastq</code> tool in cooperation with <code>multiple-mutate</code>. The tool <code>filter_fastq</code> enables you to apply transformations to existing fastq files. To this effect, <code>filter_fastq</code> extracts one part out of a FASTQ file (read name, nucleotide string or quality string) and feeds it into another sub-program specified between two <code>@</code>-signs. The sub-program is required to take lines of text as input and return the same number of lines on standard output. The output of the sub-program is then placed into the output fastq file. By combining <code>filter_fastq</code> and <code>multiple_mutate</code>, the tool which applies mutations to strings of nucleotides, a FASTQ file can be mutated. The <code>--seed</code> can be set to an arbitrary value to generate a reproducable result.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/filter_fastq</span> --nucleotide \
  @ scripts/multiple_mutate --seed 123 input/mut-tables/mut.tab @ \
  <span class="op">&lt;</span> data/2/volpertinger.fastq \
  <span class="op">&gt;</span> data/3/volpertinger_mut.fastq</code></pre></div>
<pre class="output"><code></code></pre>
<p>Note how the nucleotide strings of the output FASTQ file now carry mutations (lower-case letters):</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">head</span> data/3/volpertinger_mut.fastq</code></pre></div>
<pre class="output"><code>@volpertinger_1
cTcTAaAAGATAgTAGCCAATCAGATCCA
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@volpertinger_2
TCTAcTTAATAgTTTCTCCC
+
FFFFFFFFFFFFFFFFFFFF
@volpertinger_3
TcGgACTTTATCTTCCGccGCTCAAC</code></pre>
<p>The <code>filter_fastq.py</code> script enables you to apply an arbitrary script or program on just one part of a FASTQ file (ID line, nucleotide line, quality line). The used script must accept the respective part on standard input and print the modified version on standard output. The modified FASTQ file is assembled by <code>filter_fastq.py</code> from the output of its children scripts and printed on standard output.</p>
<p>On the <code>filter_fastq.py</code> call, the @ sign serves as a sentinel character, which determines start and end of the sub-program's command line. It can also be any arbitrary other character, as long as it doesn't occur inside the child script's command line but only at the beginning and the end.</p>
<h2 id="combining-endogenous-and-non-endogenous-reads">Combining endogenous and non-endogenous reads</h2>
<p>In this example, the endogenous reads from <em>volpertinger</em> undergo simulated mutation and damage prior to mapping, while the contaminant reads from <em>R. etli</em> do not.</p>
<p>Therefore only now, after applying mutations to our <em>volpertinger</em> reads, is the time to combine the mutated sample reads and the contaminant reads generated in the last section to one file. For this purpose, the UNIX tool <code>cat</code> is used:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">cat</span> data/3/volpertinger_mut.fastq data/2/retli.fastq \
    <span class="op">&gt;</span> data/3/all.fastq</code></pre></div>
<pre class="output"><code></code></pre>
<h1 id="advanced-read-mutation">Advanced read mutation</h1>
<p>The following two sections deal with advanced means of determining read mutation probabilities. They are not used in this tutorial, but may be used in custom workflows.</p>
<p>The subsequent section deals with how to read out mapDamage output to determine the read mutation probability parameters.</p>
<p>The section following that deals with how to create multiple distinct sets of read mutation probabilty parameters. By that means, different read sets can be created which were subjected to varying degrees of mutations. Therefore, mapper performance can be evaluated for different severities of read mutations.</p>
<h2 id="obtaining-mutation-rates-from-mapdamage">Obtaining mutation rates from mapDamage</h2>
<p>Damage patterns from mapDamage can be converted into a table with mutation parameters by least-squares fitting. For this purpose, the mapDamage output file <code>misincorporation.txt</code> is needed.</p>
<p>The following command fits a geometric distribution to mapDamage data, shows the derived parameters and plots the data with the fitted curve.</p>
<p>The <code>cut</code> command is used only to limit the output to a width acceptable for this manual. Use the bash <code>&gt;</code> redirection operator to write this output into a file suitable for <code>multiple_mutate.py</code>.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/mapdamage2geomparam</span> \
    --fit-plots fig/fit.png \
    input/mapdamage/misincorporation.txt <span class="kw">|</span> <span class="kw">\</span>
    <span class="fu">cut</span> -f1-6 <span class="kw">|</span> <span class="kw">\</span>
    <span class="ex">column</span> -t</code></pre></div>
<pre class="output"><code>strand  from  to  factor     geom_prob  intercept
3p      A     C   0.002594   0.1289     0.0003363
3p      A     G   0.005872   0.3972     0.001542
3p      A     T   0.001393   0.7516     0.0006354
3p      C     A   0.02344    0.8742     0.0007693
3p      C     G   0.004381   0.4866     0.0006858
3p      C     T   0.09154    0.566      0.003878
3p      G     A   0.02789    0.8881     0.00169
3p      G     C   2.156e-07  0.1        0.001068
3p      G     T   0.01546    0.1008     0.004605
3p      T     A   0.01236    0.6896     0.0005899
3p      T     C   9.547e-07  0.1        0.001878
3p      T     G   0.005273   0.3866     0.000331
5p      A     C   0.0003257  1          NA
5p      A     G   0.003505   0.5183     NA
5p      A     T   7.02e-07   0.1        NA
5p      C     A   0.002157   0.8335     NA
5p      C     G   0.0009197  0.6883     NA
5p      C     T   0.09852    0.9541     NA
5p      G     A   0.002752   0.6784     NA
5p      G     C   4.612e-08  0.1        NA
5p      G     T   3.004e-07  0.1        NA
5p      T     A   0.001127   0.9989     NA
5p      T     C   2.849e-07  0.1        NA
5p      T     G   0.0007032  0.8024     NA</code></pre>
<p>For each base exchange, the position-independent parameter <code>intercept</code> is only listed for one strand end.</p>
<p>This is the generated plot, which shows the mutation frequencies obtained from mapDamage (black dots) and the fitted mutation model (blue lines):</p>
<p><img src="fig/fit.png" width=600/></p>
<h2 id="generating-multiple-damage-patterns-using-a-parameter-table">Generating multiple damage patterns using a parameter table</h2>
<p>Sometimes, multiple damage patterns need to be compared. There is a possibility to generate the <code>multiple_mutate</code> input files from one table which lists all different values of the different mutation parameters.</p>
<p>This approach will be seen again later, where there is a possibility to generate many short read mapper calls from exactly the same kind of parameter table. You can therefore generate appropriate input files as well as appropriate short read mapper calls out of only one table which lists all the parameters.</p>
<p>The <code>fill_template.py</code> script expects a table, where each row is used to fill a prespecified template with values.</p>
<p>For example, if a template is written which looks like this:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">cp</span> input/mut-tmpl/mut-tmpl data/3
<span class="ex">column</span> -t data/3/mut-tmpl</code></pre></div>
<pre class="output"><code>strand  from  to  factor  geom_prob  intercept
5       C     T   {fac}   {geom}     0
3       G     A   {fac}   {geom}     0
3       *     *   0       0          {all_intercept}</code></pre>
<p>... and another table is written which lists several values for the expressions in braces <code>{...}</code> from the previous file:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">cp</span> input/mut-tmpl/tab data/3/mut-tab
<span class="ex">column</span> -t data/3/mut-tab</code></pre></div>
<pre class="output"><code>fac  geom  all_intercept
0    0.1   0
0    0.1   0.3
0.5  0.1   0
0.5  0.1   0.3</code></pre>
<p>... several files can be generated, where the brace-expressions are replaced by the values of the second table, using the <code>fill_template</code> script:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/fill_template</span> \
    data/3/mut-tmpl \
    <span class="op">&lt;</span> data/3/mut-tab</code></pre></div>
<pre class="output"><code>strand  from    to  factor  geom_prob   intercept
5   C   T   0   0.1 0
3   G   A   0   0.1 0
3   *   *   0   0   0

strand  from    to  factor  geom_prob   intercept
5   C   T   0   0.1 0
3   G   A   0   0.1 0
3   *   *   0   0   0.3

strand  from    to  factor  geom_prob   intercept
5   C   T   0.5 0.1 0
3   G   A   0.5 0.1 0
3   *   *   0   0   0

strand  from    to  factor  geom_prob   intercept
5   C   T   0.5 0.1 0
3   G   A   0.5 0.1 0
3   *   *   0   0   0.3
</code></pre>
<p>To write each table into its own file, the <code>--output</code> switch of the <code>fill_template</code> script can be used. The argument of <code>--output</code> can (and should!) contain column names of the table, again enclosed in braces <code>{...}</code>. This creates multiple files, where the braced expressions in the file names are replaced in the same way as they are in the table contents.</p>
<p>We will now write each of the tables shown above to its own file. We want to name the files using a counting number, but our input table doesn't yet contain a column with that counter. Therefore we must first add one.</p>
<p>The mentioned-above table can be prepended with an index column:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/index_column</span> --inplace data/3/mut-tab

<span class="fu">head</span> data/3/mut-tab <span class="kw">|</span> <span class="ex">column</span> -t</code></pre></div>
<pre class="output"><code>index  fac  geom  all_intercept
0      0    0.1   0
1      0    0.1   0.3
2      0.5  0.1   0
3      0.5  0.1   0.3</code></pre>
<p>Now, each output of <code>fill_template.py</code> can be written to its own output file, using the information from the newly-generated <code>index</code> column:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/fill_template</span> \
    --output <span class="st">&quot;data/3/{index}_filled&quot;</span> \
    data/3/mut-tmpl \
    <span class="op">&lt;</span> data/3/mut-tab

<span class="co"># Show all the generated files</span>
<span class="kw">for</span> <span class="ex">f</span> in data/3/*_filled<span class="kw">;</span> <span class="kw">do</span>
    <span class="bu">echo</span> <span class="st">&quot; === </span><span class="va">$f</span><span class="st"> === &quot;</span>
    <span class="ex">column</span> -t <span class="va">$f</span>
<span class="kw">done</span></code></pre></div>
<pre class="output"><code> === data/3/0_filled ===
strand  from  to  factor  geom_prob  intercept
5       C     T   0       0.1        0
3       G     A   0       0.1        0
3       *     *   0       0          0
 === data/3/1_filled ===
strand  from  to  factor  geom_prob  intercept
5       C     T   0       0.1        0
3       G     A   0       0.1        0
3       *     *   0       0          0.3
 === data/3/2_filled ===
strand  from  to  factor  geom_prob  intercept
5       C     T   0.5     0.1        0
3       G     A   0.5     0.1        0
3       *     *   0       0          0
 === data/3/3_filled ===
strand  from  to  factor  geom_prob  intercept
5       C     T   0.5     0.1        0
3       G     A   0.5     0.1        0
3       *     *   0       0          0.3</code></pre>
<p>If several combinations of mutation parameters shall be tested, <code>cross_tab.py</code> can be used to generate a table like <code>data/3/mut-tab</code> from predefined parameter values, like described with mapper parameters in the next chapter.</p>
</div>
<div class="contentbox" id="navlinksbox">
<div>
<a href="02_read-generation.html">&lt;&nbsp;Prev</a>
&nbsp;|&nbsp;<a href="#top">Go to top</a>&nbsp;|&nbsp;
<a href="04_mapping.html">Next&nbsp;&gt;</a>
</div>
</div>
</body>
</html>
