<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title>Writing a mapper script</title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
  <link rel="stylesheet" href="manual.css" type="text/css" />
</head>
<!-- BODY BEGIN ------------------------------>
<body>
<div class="toc contentbox">
    <span class="tocheader">Table of contents</span>
<ul>
<li><a href="index.html">Package overview</a></li>
<li><a href="00b_shell-prerequisites.html">Dependencies of the scripts</a>
<li><a href="00c_input-files.html">Input files</a>
<li><a href="01_genome-preparation.html">Genome Preparation</a>
<li><a href="02_read-generation.html">Generation of simulated read sets</a>
<li><a href="03_read-mutation.html">Mutation of reads</a>
<li><a href="04_mapping.html">Generate mapper calls</a>
<li class="tocthis"><a href="05_writing-a-mapper-script.html">Writing a mapper script</a>
<li><a href="06_sam-parsing.html">Parsing of SAM files</a>
<li><a href="07_parameter_influence.html">Comparing multiple mapper runs</a>
<li><a href="A01_workflow.html">Appendix A1 -- Tutorial script</a>
<li><a href="A02_advanced_read-generation.html">Appendix A2 -- Advanced read generation</a>
<li><a href="A03_scripts.html">Appendix A3 -- Overview of the scripts by category</a>
</ul>
</div>
<div class="content contentbox">
<div id="header">
<h1 class="title">Writing a mapper script</h1>
</div>
<!-- Local Table of contents -->
<!--Back to table of contents -->
<!-- CONTENT -------------------------------------->
<h2 id="the-mapping-script-is-an-adapter">The mapping script is an adapter</h2>
<p>TAPAS can be used with any short read mapping program. However, every mapper has its own program name or may even consist of multiple programs. It has its own command line arguments and possibly unique preparation or further processing steps until its result is in the SAM file format.</p>
<p>Therefore, in order that TAPAS can test parameter values for this mapper, an adapter must be created which is provided with the parameter values to be tested as input and which produces a SAM file of the mapping which was performed using that parameter values. The resulting SAM file will then be analysed in the following chapters.</p>
<h2 id="variables">Variables</h2>
<p>The mapping script will contain the commands needed for the mapping as they would be entered in the command line when performing a standard mapping outside of TAPAS. However, the difference is that in TAPAS, the parameter values are replaced by <strong>variables</strong>. For example, a variable named <code>k</code> looks like this: <code>${k}</code>.</p>
<p>The names of the variables which are used in the mapping script are determined by the first line of the table, which was provided to the tool <code>table2calls</code> in the previous chapter: <code>data/4/partab</code>. Additionally, all variable names which are provided to <code>table2calls</code> via the <code>--const</code> option.</p>
<p>Therefore, in the case of this example, the first line of <code>data/4/partab</code> contains the names</p>
<pre><code>runidx k n</code></pre>
<p>and the call to <code>table2calls</code> is</p>
<pre><code>scripts/table2calls --const genome &quot;input/genome/volpertinger&quot; \
                    --const reads  &quot;data/3/all.fastq&quot; \
                    --const outdir &quot;data/4&quot; \
                ...</code></pre>
<p>Therefore, we can use the variables <code>${runidx}</code>, <code>${k}</code>, <code>${n}</code>, <code>${genome}</code>, <code>${reads}</code> and <code>${outdir}</code> in the mapping script.</p>
<h2 id="writing-the-mapping-script">Writing the mapping script</h2>
<p>Each mapping script should start with these two lines. There may be no empty lines or comments before this:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">#!/bin/bash</span>
<span class="kw">set</span> <span class="ex">-ueC</span></code></pre></div>
<p>The first line identifies the file as a script which can be executed like a program, the second line makes sure that the script aborts with an error if there is a mistake like when an unknown variable name is used or when the mapper returns an error. The last letter, <code>C</code>, is a safety measure against accidential overwriting of files. In the case of mappers whose output file is determined via the <code>&gt;</code> operator, like in <code>bwa aln .... &gt; out.sam</code>, it protects against accidential overwriting of output files due to a programming error. If the command does not contain the <code>&gt;</code> character, e.g. <code>other_mapper --output out.sam</code>, this can provide no safety.</p>
<p>Following this, there may be a line which saves all status and error messages which are produced during the mapping into a log file. This is optional but handy to check if a mapping run was completed correctly or if it produced any errors:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="bu">exec</span> <span class="op">2&gt;</span><span class="st">&quot;</span><span class="va">${outdir}</span><span class="st">/</span><span class="va">${runidx}</span><span class="st">.log&quot;</span> <span class="op">&gt;&amp;2</span></code></pre></div>
<p>This command belongs to the bash language. It redirects all the output which is produced during the remainder of this script into the file <code>&quot;${runidx}.log&quot;</code>, inside the folder <code>${outdir}</code>. Note that this is the first use of variables in the script. The part <code>${runidx}</code> will be replaced by one of the values in the column <code>runidx</code> in the file <code>data/4/partab</code>. This column contains a steadily rising number in each line, therefore, the log of each mapping run will be put into its own file. The <code>${outdir}</code> part will be constant every time because it will be set using <code>table2calls --const outdir &quot;...SOMEFOLDER...&quot;</code>, like it was shown in the previous chapter.</p>
<p>This is an important idea: <strong>Each mapping run should have its own set of output files</strong>. The reason is that several mapping runs can be executed in parallel by TAPAS, to save time. If all mapping runs would write into the same file in parallel, the file would only contain a messy mixture of all outputs afterwards. That's bad in the case of logfiles, but still worse in the case of mapping results!</p>
<p>Therefore, it is good practice to always include a counting column to the table which holds the parameter combinations for each run (<code>data/4/partab</code>) by using <code>add_index_column</code>, like it was done in the previous chapter.</p>
<p>Finally, it can be seen that the filename (actually, its placeholder using the <code>runidx</code> variable) is enclosed in double quotes. This is another recommended practice which is important when writing file names and file paths, because those can contain spaces: <strong>Wrap variables which might contain spaces into double quotes!</strong></p>
<p>The rest of the script is only the needed program calls to perform the mapping. Below, an example for the mapper BWA is shown:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">bwa</span> aln -n <span class="va">${n}</span> -k <span class="va">${k}</span> \
    <span class="st">&quot;</span><span class="va">${genome}</span><span class="st">&quot;</span> \
    <span class="st">&quot;</span><span class="va">${reads}</span><span class="st">&quot;</span> \
    <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${outdir}</span><span class="st">/</span><span class="va">${runidx}</span><span class="st">.sai&quot;</span>

<span class="ex">bwa</span> samse \
      <span class="st">&quot;</span><span class="va">${genome}</span><span class="st">&quot;</span> \
      <span class="st">&quot;</span><span class="va">${outdir}</span><span class="st">/</span><span class="va">${runidx}</span><span class="st">.sai&quot;</span> \
      <span class="st">&quot;</span><span class="va">${reads}</span><span class="st">&quot;</span> \
      <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${outdir}</span><span class="st">/</span><span class="va">${runidx}</span><span class="st">.sam&quot;</span></code></pre></div>
<p>Note again:</p>
<ul class="incremental">
<li>The values which follow the BWA parameters <code>-n</code> and <code>-k</code> are replaced by <strong>variables</strong></li>
<li>Locations where file paths might pop up or other variables which might turn out to yield strings with spaces should be put in <strong>double quotes</strong></li>
<li>Each run of the script should produce a <strong>unique set of output files</strong>. This can be ensured by using a counter variable like <code>runidx</code> in this case.</li>
</ul>
<h2 id="parameter-onoff-switches">Parameter on/off switches</h2>
<p>So far we've only considered parametes which follow the form <code>-p VALUE</code>, <code>--parameter VALUE</code> or similar. However, there are also parameters which are on/off switches, hence, they are either appended to the command line or they are not. There is no explicit value.</p>
<p>These parameters can be tested in TAPAS by writing the parameter itself and another, arbitrary string into the corresponding <code>*.par</code> file.</p>
<p>This is an example with the BWA <code>-L</code> switch, which enables log scaling of gap penalties instead of linearly rising gap penalty.</p>
<p><code>bwa aln GENOME READS &gt; OUTPUT.sam</code> is the standard call where this setting is disabled. <code>bwa aln -L GENOME READS &gt; OUTPUT.sam</code> is the call where this setting is enabled.</p>
<p>Write the following file which serves as an input to <code>cross_tab</code> (see previous chapter). In this example, it would go into the <code>input/mapping</code> folder and would be named <code>L.par</code>, though this is an arbitrary choice:</p>
<pre><code>L
-L
xx</code></pre>
<p>The first line determines the variable name in the mapping script, the second line shows the parameter which is used to switch on the setting and the third line shows an arbitrary string which indicates the setting is switched off.</p>
<p>In the mapping script, before the variable <code>${L}</code> is used for the first time, include the following line. Take care that you insert whitespace exactly as shown below. The strings <code>L</code> and <code>xx</code> must correspond to the first and third line of the above-mentioned <code>L.par</code> file.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">if</span><span class="bu"> [</span> <span class="va">${L}</span> <span class="ot">=</span> xx<span class="bu"> ]</span>; <span class="kw">then</span> <span class="bu">unset</span> <span class="va">L</span>; <span class="kw">fi</span></code></pre></div>
<p>Then, use the parameter in the mapping call <strong>without quotes</strong> (<code>&quot;...&quot;</code>) and <strong>append a minus (-) sign</strong>. For example, to amend the code given in the section &quot;Writing the mapping script&quot; above:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">bwa</span> aln -n <span class="va">${n}</span> -k <span class="va">${k}</span> <span class="va">${L-}</span> \
    <span class="st">&quot;</span><span class="va">${genome}</span><span class="st">&quot;</span> \
    <span class="st">&quot;</span><span class="va">${reads}</span><span class="st">&quot;</span> \
    <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${outdir}</span><span class="st">/</span><span class="va">${runidx}</span><span class="st">.sai&quot;</span>

  <span class="kw">(</span><span class="ex">...</span><span class="kw">)</span></code></pre></div>
</div>
<div class="contentbox" id="navlinksbox">
<div>
<a href="04_mapping.html">&lt;&nbsp;Prev</a>
&nbsp;|&nbsp;<a href="#top">Go to top</a>&nbsp;|&nbsp;
<a href="06_sam-parsing.html">Next&nbsp;&gt;</a>
</div>
</div>
</body>
</html>
