<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title>Appendix A2 -- Advanced read generation</title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
  </style>
  <link rel="stylesheet" href="manual.css" type="text/css" />
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>
</head>
<!-- BODY BEGIN ------------------------------>
<body>
<div class="toc contentbox">
    <span class="tocheader">Table of contents</span>
<ul>
<li><a href="index.html">Package overview</a></li>
<li><a href="00b_shell-prerequisites.html">Dependencies of the scripts</a>
<li><a href="00c_input-files.html">Input files</a>
<li><a href="01_genome-preparation.html">Genome Preparation</a>
<li><a href="02_read-generation.html">Generation of simulated read sets</a>
<li><a href="03_read-mutation.html">Mutation of reads</a>
<li><a href="04_mapping.html">Generate mapper calls</a>
<li><a href="05_writing-a-mapper-script.html">Writing a mapper script</a>
<li><a href="06_sam-parsing.html">Parsing of SAM files</a>
<li><a href="07_parameter_influence.html">Comparing multiple mapper runs</a>
<li><a href="A01_workflow.html">Appendix A1 -- Tutorial script</a>
<li class="tocthis"><a href="A02_advanced_read-generation.html">Appendix A2 -- Advanced read generation</a>
<li><a href="A03_scripts.html">Appendix A3 -- Overview of the scripts by category</a>
</ul>
</div>
<div class="content contentbox">
<div id="header">
<h1 class="title">Appendix A2 -- Advanced read generation</h1>
</div>
<!-- Local Table of contents -->
<!--Back to table of contents -->
<!-- CONTENT -------------------------------------->
<p>This more verbose but also more flexible way of generating artificial reads is currently needed if you want to</p>
<ul class="incremental">
<li>Use customised base quality strings in the output FASTQ file,</li>
<li>Apply indels or other alterations to the nucleotide string which also affect the read quality string.</li>
</ul>
<p>The output FASTQ file will be constructed by interleaving the contents of three files, containing the read names, the nucleotide strings and the quality strings, respectively. These files will be generated in the following.</p>
<p>Additionally, a table containing the read names with the true read positions will be created, to evaluate later on whether a read was mapped correctly.</p>
<p>The <code>uniform</code> script can be used to sample reads from a reference genome. The command is used like already explained in the chapter about <a href="02_read-generation.html">simplified read generation</a>. The difference is that in this case it is called without the <code>--output-fastq</code> switch. Instead, the files needed for creating the read FASTQ file are created manually and merged to the FASTQ format using <code>synth_fastq</code>.</p>
<p>This manual shows an example, arbitrary custom manipulation steps can be performed in between or instead these steps.</p>
<p>Create the subfolder which holds the generated files:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">mkdir</span> data/2e</code></pre></div>
<pre class="output"><code></code></pre>
<h2 id="generating-artificial-reads-from-the-reference">Generating artificial reads from the reference</h2>
<p>Using <code>uniform</code> without the <code>--output-fastq</code> switch produces a text table output with four columns:</p>
<ol class="incremental" style="list-style-type: decimal">
<li>The FASTA record name (e.g. chromosome) where the read originated.</li>
<li>1-based base number of the reads' first base.</li>
<li>1-based base number of the reads' last base.</li>
<li>The nucleotide sequence of the artificial read.</li>
</ol>
<p>Example: Extract 25 reads, with minimum length 20 base pairs, where half of the reads longer than 25 bp. The value 1234 is used to initialize the random number generator. To obtain a reproducible output, this parameter can be used with an arbitrary integer. If omitted, different reads are produced on every call.</p>
<p>We will generate sample reads from our <code>volpertinger</code> sample genome. Execute the following script to generate random nucleotide sequences:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/uniform</span> input/genome/volpertinger.fasta \
    --seed 1234\
    25 20 5  \
    <span class="op">&gt;</span> data/2e/volpertinger.reads</code></pre></div>
<pre class="output"><code></code></pre>
<p>The resulting file looks like this:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">head</span> data/2e/volpertinger.reads <span class="kw">|</span> <span class="ex">column</span> -t</code></pre></div>
<pre class="output"><code>record  start  end   read
B1      2143   2171  TTCCACAAGATATTAGCCAACCAGATCCA
MT      3402   3421  TCTATTTAATAACTTCTCCC
A3      1413   1438  TTGAACTCTATCTTCCGGGGCTCAAC
A3      5689   5709  TCATGAGGGTTTCTATTTAGA
MT      3280   3304  TACGACCTCGATGTTGGATCAGGAC
A3      4936   4962  ACCCAAGGGCCCCCAGCCAGGTTGCTG
A1      1320   1341  TGGGAATAGCCCTGTGTTTATC
A3      4480   4506  AATAAACTTAAAAAAATTAGTGGCATG
X       2381   2400  ACGATGACCATCTTCTTGCG</code></pre>
<h2 id="insertions-and-deletions-indels">Insertions and deletions (indels)</h2>
<p>TAPAS contains a tool, called <code>indel</code>, which can insert and delete nucleotides randomly. The tool outputs the mutated nucleotide sequences along with a CIGAR string which describes which operation had been taken. It expects as input the list of nucleotides. Below is an example call with the nucleotides created above.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/indel</span> --in-prob 0.01  --in-exp 0.1 \
              --cigar-new     --seed 112233 \
              --col-seq read --col-cigar cigar \
    <span class="op">&lt;</span> data/2e/volpertinger.reads \
    <span class="op">&gt;</span> data/2e/volpertinger.reads.indel

<span class="fu">head</span> data/2e/volpertinger.reads.indel <span class="kw">|</span> <span class="ex">column</span> -t</code></pre></div>
<pre class="output"><code>record  start  end   read                             cigar
B1      2143   2171  TTCCACAATGGATATTAGCCAACCAGATCCA  8M2I21M
MT      3402   3421  TCTATTTAATAACTTCTCCC             20M
A3      1413   1438  TTGAACTCTATCTTCCGGGGCTCAAC       26M
A3      5689   5709  TCATGAGGGTTTCTATTTAGA            21M
MT      3280   3304  TACGAATCCTCGATGTTGGATCAGGAC      5M2I20M
A3      4936   4962  ACCCAAGGGCCCCCAGCCAGGTTGCTG      27M
A1      1320   1341  TGGGAATAGCCCTGTGTTTATC           22M
A3      4480   4506  AATAAACTTAAAAAAATTAGTGGCATG      27M
X       2381   2400  ACGATGACCATCTTCTTGCG             20M</code></pre>
<p><strong>Please note (I):</strong> If a deletion occurs at the beginning of the read, it cannot be detected by the mapper which is to be tested. In the optimal case a deletion of <span class="math inline">\(n\)</span> bp at the beginning of the read will cause the mapper to locate the read <span class="math inline">\(n\)</span> bp downstream of its true position. For an insertion, the read location reported by the mapper depends on whether the mapper matches the inserted base pairs to the reference or not. Therefore, it might be warranted to allow for some tolerance when determining whether a read is correctly mapped or remove the affected reads from the read set. For example, <code>awk</code> and its regular expression matching capabilities (<code>... ~ /.../</code>) can be used to this effect. The following call removes lines (=reads) in the input table (<code>volpertinger.reads.indel</code>) where the CIGAR string (column 6 = <code>$6</code>) does not start with a match (<code>^[0-9]+M</code>) and end with a match (<code>[0-9]+M$</code>). See tutorials on regular expressions and the SAM Specification (CIGAR strings) for more information.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">awk</span> <span class="st">&#39;$6 ~ /^[0-9]+M/  &amp;&amp;  $6 ~ /[0-9]+M$/&#39;</span> \
    data/2e/volpertinger.reads.indel
    <span class="op">&gt;</span> <span class="ex">data/2e/volpertinger.reads.indel.filtered</span></code></pre></div>
<p><strong>Please note (II):</strong> In the published pilot study version 1.1 of TAPAS was used, which did not have the indel feature implemented. The current version is not yet been tested on an in vivo dataset. We urge caution with the use of the indel feature. Should erroneous behaviour occur despite our efforts and care to produce correcly working software, a short notice to the authors is much appreciated.</p>
<p>The usage of <code>indel</code> shown above creates a new file to facilitate the demonstration of how the presented tools change the data. However, it is also possible to directly combine the read generation and indel insertion using the UNIX pipe:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/uniform</span> input/genome/volpertinger.fasta \
    --quiet --seed 1234  25 20 5  \
 <span class="kw">|</span> <span class="ex">scripts/indel</span> --in-prob 0.01 --in-exp 0.1 \
                 --cigar-new --seed 112233 \
                 --col-seq read --col-cigar cigar \
    <span class="op">&gt;</span> data/2e/volpertinger.reads.indel</code></pre></div>
<pre class="output"><code></code></pre>
<p>The result is the same as with the above approach using the intermediate files:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">head</span> data/2e/volpertinger.reads.indel <span class="kw">|</span> <span class="ex">column</span> -t</code></pre></div>
<pre class="output"><code>record  start  end   read                             cigar
B1      2143   2171  TTCCACAATGGATATTAGCCAACCAGATCCA  8M2I21M
MT      3402   3421  TCTATTTAATAACTTCTCCC             20M
A3      1413   1438  TTGAACTCTATCTTCCGGGGCTCAAC       26M
A3      5689   5709  TCATGAGGGTTTCTATTTAGA            21M
MT      3280   3304  TACGAATCCTCGATGTTGGATCAGGAC      5M2I20M
A3      4936   4962  ACCCAAGGGCCCCCAGCCAGGTTGCTG      27M
A1      1320   1341  TGGGAATAGCCCTGTGTTTATC           22M
A3      4480   4506  AATAAACTTAAAAAAATTAGTGGCATG      27M
X       2381   2400  ACGATGACCATCTTCTTGCG             20M</code></pre>
<h2 id="read-names">Read names</h2>
<p>In this tutorial we generate read names consisting of the organism name (<em>volpertinger</em>) followed by an underscore and a counting number.</p>
<p>There is a script included for adding this kind of column, which is shown in the next code example. You can as well use <code>awk</code> or whichever tool you like to accomplish this task if you need more sophisticated read names.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/index_column</span>  --prefix volpertinger_ \
                      --colname name  \
                      --inplace data/2e/volpertinger.reads.indel

<span class="fu">head</span> data/2e/volpertinger.reads.indel <span class="kw">|</span> <span class="ex">column</span> -t</code></pre></div>
<pre class="output"><code>name            record  start  end   read                             cigar
volpertinger_0  B1      2143   2171  TTCCACAATGGATATTAGCCAACCAGATCCA  8M2I21M
volpertinger_1  MT      3402   3421  TCTATTTAATAACTTCTCCC             20M
volpertinger_2  A3      1413   1438  TTGAACTCTATCTTCCGGGGCTCAAC       26M
volpertinger_3  A3      5689   5709  TCATGAGGGTTTCTATTTAGA            21M
volpertinger_4  MT      3280   3304  TACGAATCCTCGATGTTGGATCAGGAC      5M2I20M
volpertinger_5  A3      4936   4962  ACCCAAGGGCCCCCAGCCAGGTTGCTG      27M
volpertinger_6  A1      1320   1341  TGGGAATAGCCCTGTGTTTATC           22M
volpertinger_7  A3      4480   4506  AATAAACTTAAAAAAATTAGTGGCATG      27M
volpertinger_8  X       2381   2400  ACGATGACCATCTTCTTGCG             20M</code></pre>
<h2 id="splitting-the-read-table">Splitting the read table</h2>
<p>To create a FASTQ file out of the generated reads, three files must be prepared containing:</p>
<ol class="incremental" style="list-style-type: decimal">
<li>The list of read names</li>
<li>The list of nucleotide strings</li>
<li>The list of quality strings</li>
</ol>
<p>The first and the second file are generated from the file which contains the read coordinates and sequences (<code>volpertinger.coord.indel</code> in this example). The third list is generated using standard UNIX tools from the nucleotide strings.</p>
<p>First, the contents of <code>volpertinger.reads.indel</code>, generated in the previous section, are split into separate files containing the read names and the nucleotide sequences. This step can be performed using the UNIX tool <code>awk</code>. <code>(NR != 1)</code> removes the header line and <code>{print $4}</code> prints the corresponding column. The columns are counted starting with 1.</p>
<p>Generate raw read sequences:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">awk</span> <span class="st">&#39;(NR != 1){print $5}&#39;</span> \
    <span class="op">&lt;</span> data/2e/volpertinger.reads.indel \
    <span class="op">&gt;</span> data/2e/volpertinger.nucl</code></pre></div>
<pre class="output"><code></code></pre>
<p>Generate list of read names:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">awk</span> <span class="st">&#39;(NR != 1){print $1}&#39;</span> \
    <span class="op">&lt;</span> data/2e/volpertinger.reads.indel \
    <span class="op">&gt;</span> data/2e/volpertinger.i</code></pre></div>
<pre class="output"><code></code></pre>
<p>Show the result:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">head</span> data/2e/volpertinger.nucl</code></pre></div>
<pre class="output"><code>TTCCACAATGGATATTAGCCAACCAGATCCA
TCTATTTAATAACTTCTCCC
TTGAACTCTATCTTCCGGGGCTCAAC
TCATGAGGGTTTCTATTTAGA
TACGAATCCTCGATGTTGGATCAGGAC
ACCCAAGGGCCCCCAGCCAGGTTGCTG
TGGGAATAGCCCTGTGTTTATC
AATAAACTTAAAAAAATTAGTGGCATG
ACGATGACCATCTTCTTGCG
TATGGTATTTGGTACTATATGTAGTTTGTGTGTGGTG</code></pre>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">head</span> data/2e/volpertinger.i</code></pre></div>
<pre class="output"><code>volpertinger_0
volpertinger_1
volpertinger_2
volpertinger_3
volpertinger_4
volpertinger_5
volpertinger_6
volpertinger_7
volpertinger_8
volpertinger_9</code></pre>
<h2 id="quality-strings">Quality strings</h2>
<p>The following example generates strings of constant quality score for every read. This can be done by the UNIX <code>sed</code> tool, which replaces every character by an F:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="fu">sed</span> <span class="st">&#39;s/./F/g&#39;</span> \
      data/2e/volpertinger.nucl \
    <span class="op">&gt;</span> data/2e/volpertinger.q

<span class="fu">head</span> data/2e/volpertinger.q</code></pre></div>
<pre class="output"><code>FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFF
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF</code></pre>
<p>If you want to generate more elaborate quality strings, you are free to do so using whichever tools you desire. Just generate a list with as many lines as there are nucleotide strings in <code>volpertinger.nucl</code> to input them into the pipeline.</p>
<h2 id="putting-the-fastq-file-together">Putting the FASTQ file together</h2>
<p>The <code>synth_fastq</code> tool creates a FASTQ file from its components, nucleotide string, quality string and read name (ID line). If the file containing the read names is omitted, the reads are numbered sequentially.</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/synth_fastq</span> data/2e/volpertinger.nucl \
                    data/2e/volpertinger.q    \
                    data/2e/volpertinger.i    \
    <span class="op">&gt;</span> data/2e/volpertinger.fastq

<span class="fu">head</span> data/2e/volpertinger.fastq</code></pre></div>
<pre class="output"><code>@volpertinger_0
TTCCACAATGGATATTAGCCAACCAGATCCA
+
FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF
@volpertinger_1
TCTATTTAATAACTTCTCCC
+
FFFFFFFFFFFFFFFFFFFF
@volpertinger_2
TTGAACTCTATCTTCCGGGGCTCAAC</code></pre>
<h2 id="repeat-the-above-steps-to-generate-contaminant-reads">Repeat the above steps to generate contaminant reads</h2>
<p>The following commands generate some reads from a truncated <em>Rhizobium etli</em> genome, to supply reads which are not supposed to map. This way, contamination with non-endogenous reads is simulated.</p>
<p>Note that three kinds of abbreviations are used here:</p>
<ul class="incremental">
<li>No indels are applied to the exogenous reads in this example, therefore the process of splitting the tabular output into suitable files for generating the FASTQ file is performed using the <code>--output</code> switch of the tool <code>uniform</code>.</li>
<li>Terms like <code>const_{a,b,c}</code> are expanded by <code>bash</code> to <code>const_a const_b const_c</code> and can therefore be used to specify multiple paths which share some parts.</li>
<li>The temporary files (similar to <code>volpertinger.q</code> and <code>volpertinger.i</code> above) are omitted here by using <code>bash</code>'s <em>process substitution</em> (<code>&lt;(...)</code>) which uses the output of one argument instead of a file name which the other command expects.</li>
</ul>
<p>As these are only abbreviations and do not change functionality, you can also use the more verbose commands described in the previous sections of this chapter to generate <code>retli.fastq</code>.</p>
<p>Index genome and sample nucleotide sequences:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">samtools</span> faidx input/retli/retli_tr.fasta

<span class="ex">scripts/uniform</span> \
    input/retli/retli_tr.fasta \
    --seed 2345 \
    --output data/2e/retli.<span class="dt">{coord,nucl}</span> \
    25 20 5</code></pre></div>
<pre class="output"><code></code></pre>
<p>Read names:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/index_column</span>  --prefix retli_ \
                      --colname name  \
                      --inplace data/2e/retli.coord</code></pre></div>
<pre class="output"><code></code></pre>
<p>Put the FASTQ together:</p>
<ul class="incremental">
<li>Quality strings are generated without an intermediate file using <code>sed</code></li>
<li>Read names are extracted without an intermediate file using <code>awk</code></li>
</ul>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash"><span class="ex">scripts/synth_fastq</span> \
    data/2e/retli.nucl \
    <span class="op">&lt;(</span><span class="fu">sed</span> <span class="st">&#39;s/./F/g&#39;</span>           data/2e/retli.nucl<span class="op">)</span> \
    <span class="op">&lt;(</span><span class="fu">awk</span> <span class="st">&#39;(NR!=1){print $1}&#39;</span> data/2e/retli.coord<span class="op">)</span> \
    <span class="op">&gt;</span> data/2e/retli.fastq

<span class="fu">head</span> data/2e/retli.fastq</code></pre></div>
<pre class="output"><code>@retli_0
TCGTTCTTGTAGCCTTCCGGGAAG
+
FFFFFFFFFFFFFFFFFFFFFFFF
@retli_1
ACTGAAGCGCAAGCTCTGAAG
+
FFFFFFFFFFFFFFFFFFFFF
@retli_2
TGGCCGAGGGACGCTTGCGTCGTC</code></pre>
</div>
<div class="contentbox" id="navlinksbox">
<div>
<a href="A01_workflow.html">&lt;&nbsp;Prev</a>
&nbsp;|&nbsp;<a href="#top">Go to top</a>&nbsp;|&nbsp;
<a href="A03_scripts.html">Next&nbsp;&gt;</a>
</div>
</div>
</body>
</html>
